### PR TITLE
Import history, bookmarks, cookies, and passwords from Brave

### PR DESCRIPTION
**Pull request by garrettr**
Partially resolves brave/brave-browser#24.
Resolves brave/brave-browser#220.

Since Brave browser-laptop is also based on Chromium (via Electron/Muon), some (but not all) user data is stored in an identical or mostly identical manner as it is in Chrome/Chromium; therefore, the corresponding import methods are based on, and similar (but not identical) to those for `ChromeImporter`, including: `ImportHistory`, `ImportCookies`, `ImportPasswords`. 

`ImportBookmarks` was completely rewritten because browser-laptop stores bookmarks in a location and with a data structure that are totally different from Chromium's. However, it still uses the same high-level approach as `ChromeImporter::ImportBookmarks`—an in-order traversal of the bookmarks tree, because that's what `AddBookmarks` requires in order to maintain the relative order of the bookmarks once imported.

Some work is intentionally being left for follow-up issues (which I will file soon), including:

* Brave equivalent of `ChromeProfileLock`
* `BraveImporter::ImportFavicons`. Importing favicons from Brave is more complicated than it was for Chrome because Brave stores favicons in the cache along with all the other web content.

This PR also fixes some minor issues in `ChromeImporter` that were discovered during the development process.

## Submitter Checklist:

- [ ] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [ ] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.

## Test Plan:

1. `yarn start`. Go to the **Brave** menu and choose **Import Bookmarks and Settings**. Choose a Brave profile to import. Check Bookmarks, History, Cookies and Passwords to verify that they were imported correctly.
2. `yarn test brave_unit_tests`. All `BraveImporterTest.*` should pass.

This PR has been successfully manually tested on macOS, Windows 10, and Linux (Ubuntu 16.04, Brave installed via [Debian package](https://github.com/brave/browser-laptop/blob/master/docs/linuxInstall.md#debian-jessie-stretch-and-ubuntu-artful-zesty-yakkety-xenial-and-trusty-amd64)).

There are known issues with importing cookies and passwords from Brave when installed on Linux via [Snapcraft](https://github.com/brave/browser-laptop/blob/master/docs/linuxInstall.md#snapcraft) which will have to be addressed in a follow-up.

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions

### CODE REVIEW

File: utility/importer/brave_importer.h [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+#include "base/files/file_path.h"
+#include "base/macros.h"
+#include "base/nix/xdg_util.h"
+#include "base/values.h"
+#include "build/build_config.h"
+#include "chrome/common/importer/imported_bookmark_entry.h"
+#include "chrome/utility/importer/importer.h"
+
+class BraveImporter : public Importer {
```
Comment (darkdh) [inline utility/importer/brave_importer.h]: We should inherit ChromeImporter and make `Import*` virtual functions so that we don't have to duplicate codes for `ImportPasswords` and `ImportCookies`

File: test/data/import/brave-browser-laptop/README [hunk 2]
```diff
@@ -0,0 +1,9 @@
+Since Brave browser-laptop is based on Chromium (via Muon/Electron), most of the notes from the ChromeImporter test data README (`../chrome/README`) are also applicable here.
+
+# Generating test data
+
+Unlike Chrome/Chromium, Brave browser-laptop does not support multiple profiles, so if you use Brave as your regular web browser, it may not be immediately obvious how you can create a separate profile for the purpose of generating test data. I found the least obtrusive method to be setting the CHROME_USER_DATA_DIR environment variable, e.g. on macOS:
+
+    TEST_UDD=/path/to/test/user/data/dir
+    mkdir -p $TEST_UDD
+    CHROME_USER_DATA_DIR=$TEST_UDD /Applications/Brave.app/Contents/MacOS/Brave
```
[the remark this answers is not in the document]
Comment (garrettr) [reply test/data/import/brave-browser-laptop/README]: Good to know! I didn't know it was also a command line flag, will test again to verify that it works and update these docs.

[the remark this answers is not in the document]
Comment (garrettr) [reply test/data/import/brave-browser-laptop/README]: @darkdh As I recall now, I did try to use `--user-data-dir-name` at first, but unfortunately it does not seem to work as intended. It causes Brave to start with a fresh profile, but that profile's data never gets saved in the specified user data directory. After quitting Brave, the specified user data directory is empty. This does not happen when I use `CHROME_USER_DATA_DIR`. Saving the temporary state to disk is critical for my use case, since I need to extract the relevant Brave state files to use for test data.

[the remark this answers is not in the document]
Comment (darkdh) [reply test/data/import/brave-browser-laptop/README]: I see, that sounds like a bug in muon

[the remark this answers is not in the document]
Comment (garrettr) [reply test/data/import/brave-browser-laptop/README]: @darkdh Upon further testing, it looks like `--user-data-dir` has the expected behavior, while `--user-data-dir-name` does not. `--user-data-dir` is a Chromium [flag](https://peter.sh/experiments/chromium-command-line-switches/#user-data-dir), I'm not sure but it seems `--user-data-dir-name` might have been introduced in Muon. Should I update the documentation to recommend `--user-data-dir` instead of `CHROME_USER_DATA_DIR`?

[the remark this answers is not in the document]
Comment (darkdh) [reply test/data/import/brave-browser-laptop/README]: yes, please. thanks

File: utility/importer/brave_importer.cc [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+
+     if (type == "bookmark-folder") {
+      base::Value* bookmark_folder =
+        bookmark_folders_dict->FindKeyOfType(key, base::Value::Type::DICTIONARY);
+      auto& title =
+        bookmark_folder->FindKeyOfType("title", base::Value::Type::STRING)->GetString();
+
+      // Empty folders don't have a corresponding entry in bookmark_order_dict,
+      // which provides an easy way to test whether a folder is empty.
+      base::Value* bookmark_order_entry =
```
Comment (darkdh) [inline utility/importer/brave_importer.cc]: this seems weird to me because empty bookmark folder still requires order. (ex. you can rearrange their position)
Also, I verified it does exist in `bookmarkOrder` with key equals to folderId

**Comment by darkdh** [comment]
great work on `ChromeImporter::ImportBookmarks` 👍

File: patches/chrome-browser-importer-external_process_importer_client.cc.patch [hunk 4]
```diff
@@ -0,0 +1,22 @@
+diff --git a/chrome/browser/importer/external_process_importer_client.cc b/chrome/browser/importer/external_process_importer_client.cc
+index 429c16dc6a56fcd0ea1d674508c1927a387e1905..0f29ba75ffce20b1104e36b4ba1aa78fd28fc50a 100644
+--- a/chrome/browser/importer/external_process_importer_client.cc
++++ b/chrome/browser/importer/external_process_importer_client.cc
+@@ -8,6 +8,7 @@
+ 
+ #include "base/bind.h"
+ #include "base/strings/string_number_conversions.h"
++#include "brave/grit/generated_resources.h"
```
Comment (darkdh) [inline patches/chrome-browser-importer-external_process_importer_client.cc.patch]: Please rebase to master. We don't need to patch this now, 
see https://github.com/brave/brave-core/pull/179/commits/5b4838a4f32f7b38f5420287b0f547f4514aba8f

File: patches/chrome-browser-importer-importer_list.cc.patch [hunk 5]
```diff
@@ -11,11 +11,13 @@ index 3bf47fa746e2a417f89201f3e3800d8639f2e323..122e9daf552834db4d41c77a32033230
  #include "base/task_scheduler/task_traits.h"
  #include "base/threading/thread_restrictions.h"
 +#include "base/values.h"
++#include "brave/common/importer/brave_importer_utils.h"
 +#include "brave/common/importer/chrome_importer_utils.h"
++#include "brave/grit/generated_resources.h"
```
Comment (darkdh) [inline patches/chrome-browser-importer-importer_list.cc.patch]: ditto

**Comment by garrettr** [comment]
@darkdh Ready for re-review

**Comment by darkdh** [comment]
lgtm, would you mind doing a rebase? it shows conflict against master.

**Comment by garrettr** [comment]
@darkdh Rebased!

File: utility/importer/brave_importer.cc [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+  }
+  std::vector<std::unique_ptr<autofill::PasswordForm>> blacklist;
+  success = database.GetBlacklistLogins(&blacklist);
+  if (success) {
+    for (size_t i = 0; i < blacklist.size(); ++i) {
+      bridge_->SetPasswordForm(*blacklist[i].get());
+    }
+  }
+#else
+  base::FilePath prefs_path =
```
Comment (darkdh) [inline utility/importer/brave_importer.cc]: actually can we pass as argument? since it is the only deviation compared to `ChromeImporter:: ImportPasswords` so that we don't have to override it

File: patches/chrome-browser-ui-content_settings-content_setting_image_model.cc.patch [hunk 7]
```diff
@@ -1,5 +1,5 @@
 diff --git a/chrome/browser/ui/content_settings/content_setting_image_model.cc b/chrome/browser/ui/content_settings/content_setting_image_model.cc
-index 366ba05a508c8e5f77bd0a380ff229da73065819..b24ef19f81ab2c469780caab71a5953d95750d16 100644
+index 366ba05a508c8e5f77bd0a380ff229da73065819..4f159eee09eb691c796681af5e669df442aca7b7 100644
```
Comment (darkdh) [inline patches/chrome-browser-ui-content_settings-content_setting_image_model.cc.patch]: this patch seems irrelevant to this PR

File: patches/chrome-utility-importer-profile_import_impl.cc.patch [hunk 8]
```diff
@@ -19,7 +19,7 @@ index 1cbcc2fcb663c43daad9e6ad86f97e74b964aa11..2910707b7f569cb87947c81d8522f1d9
  #include "base/threading/thread.h"
  #include "base/threading/thread_task_runner_handle.h"
  #include "build/build_config.h"
-@@ -39,6 +43,16 @@ void ProfileImportImpl::StartImport(
+@@ -39,6 +43,19 @@ void ProfileImportImpl::StartImport(
```
Comment (darkdh) [inline patches/chrome-utility-importer-profile_import_impl.cc.patch]: can we subclass `ProfileImportImpl` and only overrides `StartImport`?
In overridden method do OSCrypt password hack first and then call parent's StartImport
so we only need to substitute `ProfileImportImpl` to `BraveProfileImportImpl` by preprocessor in `chrome/utility/importer/profile_import_service.cc` with only one line

Comment (darkdh) [reply patches/chrome-utility-importer-profile_import_impl.cc.patch]: There is one unrelated request but it will allow us to remove this patch completely
can we use preprocessor for `BraveExternalProcessImporterBridge` ?

File: patches/components-os_crypt-key_storage_keyring.cc.patch [hunk 9]
```diff
@@ -19,15 +19,16 @@ index af2f64b34ddb8e49d72dd0dd22ffd2418a422efb..9db6e47b35e6cac14168908a4385120f
...
+@@ -45,9 +46,19 @@ std::string KeyStorageKeyring::GetKeyImpl() {
  
    std::string password;
    gchar* password_c = nullptr;
 +  const char* application_name;
 +  base::CommandLine* command_line = base::CommandLine::ForCurrentProcess();
 +  if (command_line->HasSwitch("import-chrome")) {
 +    application_name = "chrome";
-+  } else if (command_line->HasSwitch("import-chromium")) {
++  } else if (command_line->HasSwitch("import-chromium") ||
```
Comment (darkdh) [inline patches/components-os_crypt-key_storage_keyring.cc.patch]: let's use preprocessor to get rid of this patch too

File: patches/components-os_crypt-key_storage_kwallet.cc.patch [hunk 10]
```diff
@@ -19,7 +19,8 @@ index aa4b3d4a6b5c7b1f725d6446ea5e573094ec935b..4c4c7191a57afc7a42dfa87701578bd9
 +  if (command_line->HasSwitch("import-chrome")) {
 +    folder_name = "Chrome Keys";
 +    key = "Chrome Safe Storage";
-+  } else if (command_line->HasSwitch("import-chromium")) {
++  } else if (command_line->HasSwitch("import-chromium") ||
```
Comment (darkdh) [inline patches/components-os_crypt-key_storage_kwallet.cc.patch]: ditto

File: patches/components-os_crypt-key_storage_libsecret.cc.patch [hunk 11]
```diff
@@ -28,7 +28,8 @@ index a1b5b975bc89c4891643adab17ce0b00ba2a8032..9d4be365463d2555101a96006ca4c68a
 +  base::CommandLine* command_line = base::CommandLine::ForCurrentProcess();
 +  if (command_line->HasSwitch("import-chrome")) {
 +    application_name = "chrome";
-+  } else if (command_line->HasSwitch("import-chromium")) {
++  } else if (command_line->HasSwitch("import-chromium") ||
```
Comment (darkdh) [inline patches/components-os_crypt-key_storage_libsecret.cc.patch]: ditto

File: patches/components-os_crypt-keychain_password_mac.mm.patch [hunk 12]
```diff
@@ -26,7 +26,8 @@ index 2b38db266f9aa1f4141c8649c021042ede4e5589..4b14f15e5091b251be53eff29139beaa
 +  if (command_line->HasSwitch("import-chrome")) {
 +    service_name = "Chrome Safe Storage";
 +    account_name = "Chrome";
-+  } else if (command_line->HasSwitch("import-chromium")) {
++  } else if (command_line->HasSwitch("import-chromium") ||
```
Comment (darkdh) [inline patches/components-os_crypt-keychain_password_mac.mm.patch]: ditto

**Comment by darkdh** [comment]
@garrettr could you open an issue for follow up? Thanks.
I'm gonna merge this PR first.

**Comment by bbondy** [comment]
WHOOOOOHOOOO 🎉
